### DISCUSSION TITLE
Downloading/Offline Mode Roadmap

### DISCUSSION THREAD

**Discussion by LePips** [jellyfin/Swiftfin#364] [Discussions]
The status and progress of Downloading/Offline Mode are described below. Issues will not be accepted for this feature until **Beta**. Each stage will accompany design improvements.

### Pre-Pre-Alpha
- basic downloading
- basic offline mode
- basic progress tracking
- read from local file if available for any item

### Pre-Alpha
- Series+Season categorized downloads
- more stats on download progress: size downloaded and estimated time left
- split video view models into online/offline instead of a single one, class based
- properly intended offline experience with custom views

### Alpha
- allow choosing what file downloaded instead of first listed on the item
- download external subtitles and have video player read them from Pre-Alpha video view model refactoring
- all download caches are cleared as expected

### Beta
- all implemented as expected above, get bugs from testers
- take a look at downloading alternate versions and sizes from server, if available from hls streams (not the same as choosing what file to download)
- strings localized

**Comment by ClappTV** [comment]
Would there be the possibility of putting some type of DRM or encryption so that it can only be played from the application itself? without being able to get the original file.

**Reply by LePips** [reply]
No, we can only download the original file. It isn't worth it to do anything else to it like encryption.

**Reply by ClappTV** [reply]
I have a server with tutorial videos for a school and in my case, I would not like people to be able to obtain them and share them on the internet without my consent.
Could it be reconsidered in the future? I'm sure there will be more cases like mine.

**Reply by LePips** [reply]
The only thing that we are able to do is adhere to the user permission to download media as set in the server. We can't do encryption or DRM, that's an entirely different technology of its own I can't learn due to time. The only thing that we'd be able to do would be to hide the downloads folder from the basic Files app but that can be easily circumvented. Even then, it's more appropriate to show the downloads and having a toggle on the app would completely beat the purpose.

**Reply by LordFransie** [reply]
I'd look into one of the Adobe media management tools if you want DRM. They're pricey but they do what you want.

**Reply by ClappTV** [reply]
> I'd look into one of the Adobe media management tools if you want DRM. They're pricey but they do what you want.

Could you provide a little more information on that? how could it be implemented in jellyfin?

**Reply by thornbill** [reply]
DRM of any kind is against the ideals of the Jellyfin Project and will not be implemented.

The Adobe suggestion is an alternative to using Jellyfin.

**Reply by LordFransie** [reply]
Yeah apologies if I wasn’t clear enough. You’re going to need to switch to a completely different piece of software and it’s going to be quite expensive. DRM is not a cheap tool.

**Comment by LordFransie** [comment]
Has this item made any progress?

**Reply by LePips** [reply]
Some, I had some work a while ago here https://github.com/jellyfin/Swiftfin/pull/362. Just have to work on other stuff before I get back to it.

**Comment by araheem00** [comment]
any update regarding whereabouts this is at?

**Comment by balajeek** [comment]
Just installed on testflight and i have to say the app looks awesome and fluid.  Keep the good work.
First i was looking is if i can download for offline viewing, and i can't seem to find any options, so i came here searching for and found this thread talking about Downloading.  Is download option still in development ?

Until then i will test with direct play and let you know of any issues.  Thank you.

**Comment by Sanfe** [comment]
A scheluder to download only at some hours would be a great feature. Some mobile operators let downloads unlimited at night.

**Comment by sparky1499** [comment]
Wondering if this is still in the roadmap?

**Reply by eduardoruizgr** [reply]
Are you guys still working on bringing this feature to swiftfin? Thanks a lot!

**Comment by StefanVDWeide** [comment]
It's been a few months since there was any activity in this discussion, so I wanted to ask again if there are any status updates on the downloading/offline mode. Having this available would be a great QoL improvement!

**Comment by IrateNinja** [comment]
There appears to already exist at least some of this functionality, however it is currently hidden and undocumented.

On [ConnectToServerViewModel.swift#L33](https://github.com/jellyfin/Swiftfin/blob/main/Shared/ViewModels/ConnectToServerViewModel.swift#L33) there is code that enable an otherwise hidden/undocumented defaults boolean `.Experimental.downloads` when the SHA256 hash of the server url matches a hardcoded value. When this defaults value is true, it adds a download button to the media page.

I've created a branch that adds a switch to the experimental section of the app settings to enable this for other servers that do not match the hard coded value. https://github.com/IrateNinja/Swiftfin/tree/feature/add-experimental-downloads-setting

**Reply by LePips** [reply]
To be clear, I added that for myself while I was traveling and didn't want to lose the work during my large PR. I highly do not recommend its usage unless you know _exactly_ what it does.

**Reply by IrateNinja** [reply]
Understood. I created the branch for the same reason you implemented this in the first place, wanted to be able to access media while traveling and in offline environments. 

I came with the intention of contributing some time towards this goal (supporting downloading/offline media), only to discover you had done some/most of/a lot of the work already. I did see #362, however, the changes involving `.Experimental.downloads` seem to be part of a *huge* commit that touches just about everything in the code base. 

Can I get a breakdown of what work has been done, what needs testing, and what is still to do?

**Comment by muava12** [comment]
any update?

**Comment by dm3ch** [comment]
Are there any formalized things, that newbies can try to do, to help to implement this?

P.S. I'm not sure if there're any Swifters in this discussion, but maybe there's something even beginners can help with

**Comment by jordibeen** [comment]
Any updates regarding this? I'm trying to adopt Jellyfin for viewing and managing home media, but I'm currently unable to download photos to iOS devices. It is possible using Safari as a workaround, but unfortunately the browser's image viewer is a bit buggy on mobile. I think this feature would be great for the native iOS application as well!

**Comment by radaiko** [comment]
This feature would really help to get more Plex users to Jellyfin.

I travel a lot and could test this as an experimental feature!

**Reply by C2gl** [reply]
Same, im looking forward for this feature, as for me its the final feature required to have jellyfin be my main media content source xd

**Comment by tim-wingspan** [comment]
up!

**Comment by shinobinakamoto** [comment]
Any update on this feature request?

Currently sitting in an airport listening to music via Jellyfin whilst thinking about what I will do aboard the plane...

**Comment by fidoriel** [comment]
Initially, I did not want to comment to make the situation worse. I also would love to see the feature soon. Writing this message, I know I am contributing to make the situation worse. I subscribed to receive PROGRESS on this feature and not messages from people asking for news or telling how much they are benefiting from this feature, how much they are anticipating it.

I am not a contributor of any Jellyfin related project, nor do I sell software in this space.
All Jellyfin creators do this in their spare time. This is visible sometimes, for example in this feature. You can use FinAmp (FOSS, https://apps.apple.com/de/app/finamp/id1574922594) for offline Music and Infuse (paid, https://firecore.com/infuse) for offline Video. Yes, Infuse is paid, so they can have some Devs for doing that. Jellyfin does not. I do only know a little bit about Swift, so I cannot help. There is also another issue wanting for download to be implemented in Jellyfin expo, the JS based Jellyfin app (https://github.com/jellyfin/jellyfin-expo/issues/372). I will use the app that delivers this feature first.

But PLEASE stop asking for progress. It is finished when it is finished. The devs are doing this for free in their spare time. Therefor, they are also developing the features they want to use first. This spams the whole discussion. They will not finish sooner if they are receiving more messages, asking for it. Also, not if they are receiving 3 Messages within 3 weeks. Asking after some months is fine, but within weeks does not help anyone. I am annoyed despite not being a contributor/assignee to this issue/project, just someone interested, so it is in my best interest to not annoy the contributors from their work I am benefitting. Please post only constructive messages, contribute code, test something etc. (This includes also this message, sorry for that). I do want to receive an email from GitHub when something important happens, to see progress and not click it away, so I can identify when I can contribute something useful or use it if the issue closes. Not when the next person cannot listen to music or watch their TV show on a plain. Please use the arrow up to show your interest/consent with a Message.

Sorry for this message again.
Thank you to all the great contributors making this possible. I will only show up again if I can help with my knowledge or skill set in a productive way.

**Reply by ult1m8** [reply]
Just a quick note to say that I work in software development (not Jellyfin) and client feedback is exactly how we determine what to work on next... otherwise you are working on what *you* think the client wants instead of what the client actually wants.

Also I wanted to note, as someone that has been keeping an eye on Jellyfin for a long time, this feature is the last thing that is keeping me from moving from Plex.

**Reply by adiberk** [reply]
I work in the industry as well and I couldn’t agree more. In general open source work relies on feature requests and bug reporting to help prioritize what to work on next.
Having more people comment can help ensure it stays on their radar.
Also just to add. I think most people, when commenting or reporting issues/feature requests do so knowing that open source contributors are incredible and are heroes. But I appreciate your sentiment and the wish to convey gratitude! It is definitely an important reminder!

**Reply by fidoriel** [reply]
I do also work in the industry, therefor I know that money can make things faster, and it is possible to sponsor a Ticket or contribute if you know the stack.

There is a new app that does support downloads: https://github.com/fredrikburmester/streamyfin
there is also a promising PR in the pipe, the contributor just needs some time to make it shiny: https://github.com/jellyfin/Swiftfin/pull/1065. Because he seems to have a live, he is on holiday, therefor it may take some time.

**Reply by EugenMayer** [reply]
infuse pro does support offline downloads with Jellyfin as backen, works far better then Plex, which has been broken all the time (10+ years plex user myself and switched recently)

**Comment by chielboogaard** [comment]
Would still love this feature to be a part of Swiftfin or Jellyfin apps. I have a setup that is non-public facing so getting my podcasts, music and movies offline would be amazing!

**Comment by jackcrane** [comment]
I built a dead-simple really bare bones (ios only) app that allows downloading Jellyfin content. It is completely free, I don't collect any data, I just built this because I wanted it, and figured why not share it with the world.

https://apps.apple.com/us/app/sunkfin/id6743207349
